core/eth/log_broadcaster_test.go

```diff
@@ -0,0 +1,260 @@
+
+	for _, log := range addr1SentLogs {
+		chRawLogs <- log
+	}
+	for _, log := range addr2SentLogs {
+		chRawLogs <- log
+	}
+
+	require.Eventually(t, func() bool { return len(addr1Logs1) == len(addr1SentLogs) }, time.Second, 10*time.Millisecond)
+	require.Eventually(t, func() bool { return len(addr1Logs2) == len(addr1SentLogs) }, time.Second, 10*time.Millisecond)
+	require.Eventually(t, func() bool { return len(addr2Logs1) == len(addr2SentLogs) }, time.Second, 10*time.Millisecond)
+	require.Eventually(t, func() bool { return len(addr2Logs2) == len(addr2SentLogs) }, time.Second, 10*time.Millisecond)
+
+	for i := range addr1SentLogs {
+		require.Equal(t, addr1SentLogs[i], addr1Logs1[i])
+		require.Equal(t, addr1SentLogs[i], addr1Logs2[i])
+	}
+	for i := range addr2SentLogs {
+		require.Equal(t, addr2SentLogs[i], addr2Logs1[i])
+		require.Equal(t, addr2SentLogs[i], addr2Logs2[i])
+	}
+
+	ethClient.AssertExpectations(t)
+}
+
+func TestLogBroadcaster_SkipsOldLogs(t *testing.T) {
+	ethClient := new(mocks.Client)
+	sub := new(mocks.Subscription)
+
+	chchRawLogs := make(chan chan<- eth.Log, 1)
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) { chchRawLogs <- args.Get(0).(chan<- eth.Log) }).
+		Return(sub, nil).
+		Once()
+
+	sub.On("Unsubscribe").Return()
+
+	lb := eth.NewLogBroadcaster(ethClient)
+	lb.Start()
+
+	addr := cltest.NewAddress()
+	logs := []eth.Log{
+		{Address: addr, BlockNumber: 0, Index: 0},
+		{Address: addr, BlockNumber: 0, Index: 1},
+		{Address: addr, BlockNumber: 0, Index: 2},
+		{Address: addr, BlockNumber: 1, Index: 0},
+		{Address: addr, BlockNumber: 1, Index: 1},
+		{Address: addr, BlockNumber: 1, Index: 2},
+		{Address: addr, BlockNumber: 2, Index: 0},
+		{Address: addr, BlockNumber: 2, Index: 1},
+		{Address: addr, BlockNumber: 2, Index: 2},
+	}
+
+	var recvd []interface{}
+	lb.Register(addr, eth.NewFuncLogListener(func(log interface{}, err error) {
+		require.NoError(t, err)
+		recvd = append(recvd, log)
+	}))
+
+	chRawLogs := <-chchRawLogs
+
+	for i := 0; i < len(logs); i++ {
+		for _, log := range logs[0 : i+1] {
+			chRawLogs <- log
+		}
+	}
+
+	lb.Stop() // This should ensure that all sending is complete
+
+	require.Len(t, recvd, len(logs))
+	for i := range recvd {
+		require.Equal(t, recvd[i], logs[i])
+	}
+
+	ethClient.AssertExpectations(t)
+}
+
+func TestLogBroadcaster_ResubscribesToMostRecentlySeenBlock(t *testing.T) {
+	const expectedBlock = 3
+
+	ethClient := new(mocks.Client)
+	sub := new(mocks.Subscription)
+
+	chchRawLogs := make(chan chan<- eth.Log, 1)
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			chchRawLogs <- args.Get(0).(chan<- eth.Log)
+		}).
+		Return(sub, nil).
+		Once()
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			query := args.Get(1).(ethereum.FilterQuery)
+			require.Equal(t, big.NewInt(expectedBlock), query.FromBlock)
```

Comment:
Need to figure out how to test the contents of `query.Addresses` despite the fact that its contents are randomly ordered.